example3/readme.md

```diff
@@ -0,0 +1,719 @@
+suggest a class is better. On usability, all else being equal, .NET recommends using concrete types that a developer can
+create via the 'new' operator. Largely this is to minimize the number of concepts or abstractions that a developer needs
+to understand to use the API. For performance calls through interfaces are harder for compilers to devirtualize and
+inline which can create small (nanosecond scale) penalties for using an interface. Some metric scenarios are expected
+to be performance critical so there is no reason for us to potentially worsen performance if we get no benefit from it.
+
+## Do users create a new Metric directly with 'new' or is the creation indirected through a factory (or both)? 
+
+Proposed answer: Use 'new' (and maybe also have a factory option purely to harmonize with the OT spec?)
+
+#### 'new' option
+````C#
+Counter c = new Counter("hats-sold");
+Gauge g = new Gauge("temperature");
+````
+
+#### factory option
+There are different reasons a factory might be useful and the naming would probably change depending
+on why we wanted it. Current OT spec uses factories to provide an extensibility point to substitute
+different implementations of the metric objects. Another use of a factory is to provide an object that
+represents a common set of metadata or configuration information that can be shared by a group of metrics.
+A 3rd use of a factory is to provide a grouping mechanism so that metric data subscribers can register
+their interest in a group of metrics rather than selecting individual metrics. A 4th use is to allow
+metric creation functions to potentially reuse existing objects, such as deduplicating based on name or
+recycling objects with an allocation pool. A 5th use is to match past expectations for users that used
+other factory based metric or telemetry APIs such as Prometheus.NET, Micrometer, 
+````C#
+// the arguments and likely type name depend on why we are using a factory
+MetricFactory factory = new MetricFactory(...); 
+Counter c = factory.CreateCounter("hats-sold");
+Gauge g = factory.CreateGauge("temperature");
+````
+
+#### both 'new' and factory option
+This option only works for a subset of the reasons we might want a factory. It does not work if the
+factory was intended to pick the concrete type that instantiates the metric or if the factory is
+intended to reuse objects, the other uses above are fine.
+````C#
+// the arguments and likely type name depend on why we are using a factory
+MetricFactory factory = new MetricFactory(...); 
+
+// dev could write this
+Counter c = factory.CreateCounter("hats-sold");
+Gauge g = factory.CreateGauge("temperature");
+
+// OR this
+Counter c = new Counter(factory, "hats-sold");
+Guage g = new Gauge(factory, "temperature");
+
+// OR potentially still this if there is a default global factory
+Counter c = new Counter("hats-sold");
+Guage g = new Gauge("temperature");

```

Comment:
Comparing with tracing:
ActivitySource source = new ActivitySource("name", "ver1.0");
Activity myActivity = source.Create/StartActivity("myactivity", Kind.Client);
myActivity?.SetTag()...
myActivity?.Stop();

The following matches the above model most?
MetricSource source = new MetricSource("name", "ver1.0");
var counter = source.CreateCounter("mycounter"); // if no listener, null returned
counter?.Increment(100, labels);

Reply:
structurally it is similar, semantically there are some important differences:
a) If a user created the MetricSource and the counter before the SDK registered or before the SDK started listening to a particular counter, returning null at that point means they will never log any measurements on that metric for the lifetime of the app. 
b) Activities are created per-request so there could be 1M activities/sec in a high perf web app. Counters are created once, intended to live for the lifetime of the app, and there should be no counters getting created at steady state. This means there could be lots of allocation caused by steady-state Activity usage but none for steady-state counter usage. Returning null was a perf optimization to compensate for that high allocation rate.

Reply:
Thanks - this clarifies a lot!

Reply:
Would you imagine having something like counter.HasListeners() or equivalent? I understand if there are no listeners, these are no-ops, but to help users optimize retrieving some non-trivial label values, a check like this would help. 
Its more of a Otel spec question, as knowing if there is a listener or not alone is not likely sufficient. Probably need more info like which dimensions are doing to be used by the listeners.